### ISSUE TITLE
Support for `required`

### ISSUE THREAD

**Issue by osterman** [hairyhenderson/gomplate#310]
## what
A `required` function that allows you to declare a particular values entry as required for template rendering. If the value is empty, the template rendering will fail with a user submitted error message.

This function is available in `helm`. 

## why
* Sometimes we need to ensure user has defined specific environment variables or the rendered template is invalid. 
* Outputting helpful messages will make this easier

## example
```
{{- required "A valid API_KEY entry required!" .Env.API_KEY -}}
```

## references:
- https://github.com/kubernetes/helm/blob/master/docs/charts_tips_and_tricks.md#know-your-template-functions

**Comment by nicolaiskogheim** [comment]
TL;DR: Why the first argument?

----------------

I think the `required` function is useful in helm charts, but I always think it's so verbose to write ".My.Key is required" every time. In your example you write `"A valid API_KEY entry required!"`, which may seem like a case for letting users define their own messages, but doing anything else than checking that the key is defined (and maybe has a non-zero length) is, generally speaking, quite hard. What I mean is that you won't know whether the `API_KEY` that is passed is valid, so the error message (or it's absence, rather, after supplying a value) is even slightly misleading.

I'm guessing that the message parameter exists because it's just easier to not have to do reflection on the required key, but other than that I don't really see the point. Anything but checking the presence of a key is out of scope for `required`, IMO. 
Since it's the library implementing `require` that decides the semantics, i.e. whether a variable satisfies `required` when it's the empty string, I think the library should define the error message as well.
Then, if the semantics change one day, the message can change as well (breaking semantics is bad, but it is near impossible if messages are user-defined). This last point shouldn't decide anything, but good qualities tend to come for free with good design in my experience, so I thought it worth mentioning.

And I don't think documentation should live in the error messages, btw, but some other place that makes sense for whatever project the template is used in. My argument is that documentation then would be split, and if it isn't split, then you only have documentation available when you use the template wrong, which I think is worse.

My $0.02

**Comment by hairyhenderson** [comment]
Thanks for your solid feedback, @nicolaiskogheim, and thanks @osterman for logging this - I know it's been a month, sorry for sitting on this - I haven't been totally ignoring it 😉

I'd like to see gomplate useful for the kind of user that would really benefit from this sort of feature. That said, I'm wondering what the _exact_ use-case is for this, and whether a targeted `env.Required` function would do the trick. Basically, while `helm` and other tools that use `text/template` put all manner of things in the context, the only thing that gomplate puts in there (at the moment) is `.Env` - i.e. a map of environment variables. So, `required` would only be useful in reality for environment variables, which was alluded to.

Maybe usage such as this would make sense:

```console
$ unset FOO
$ unset BAR
$ gomplate -i '{{ env.Required "FOO" "BAR" }}'
Error: The following environment variables are required, but were not provided: FOO, BAR
```

Another case of missing data can happen with data sources, and there's already the `datasourceExists` and `datasourceReachable` functions to help with this, though using these to _require_ that these datasources exist would currently mean wrapping the whole template in an `if`/`end` block, which is not desirable.

For that case, a simple `fail` function might do the trick:

```console
$ gomplate -i '{{ if not (datasourceExists "data") }}{{ fail "optional message" }}{{ end }}'
Error: ... optional 
```

On the _other_ hand, maybe what's really needed _is_ a `required` function, but with different semantics - basically it takes a `bool` (and maybe an optional string for message), and mix that in with a new `env.IsSet` function, and you have:

```
$ unset FOO
$ gomplate -i '{{ required (env.IsSet "FOO") }}'
...
```

Maybe that's even less friendly to the template author 🤔

**Comment by nicolaiskogheim** [comment]
I get the feeling that you're not familiar with how `required` works with helm, @hairyhenderson, and since @osterman (and myself tbh) may have made an assumption or two, I just want to point out something:

Templating of values in helm looks like this:

```yaml
myVar: {{ .Values.myVar }}
```

You can choose to pass that variable through a function `required`, which just returns the value, or fails if it isn't set:

```yaml
myVar: {{ required ".Values.myVar is required!" .Values.myVar }}
```

Now, what you suggest looks more akin to a "guard", placed maybe in the start of the file, such that one can do some light testing on whether some requirements are met, and fail if it isn't.

Or are we on the same track? :)

---------

While we're at it I would like to consider an option. It's built upon the premise that template authors, more often than not, expect that variables have values. That they create templates, happy path first like we write our programs, where everything has a proper value, and then they go "oh, shoot, what if there's no value and the thing that reads the rendered template goes haywire, blurts error messages at the user who would not understand them because they lack context or knowledge of that tool -- or worse, the thing may run silently as if everything works". And THEN they sprinkle their template, the `Makefile` that runs `gomplate`, or something else with `require` logic, leaving it slightly harder to read.

Take [elm](http://elm-lang.org/) for instance, a programming language people love, because, among other things of course, it let's you know every single time something is wrong. There is very few surprises, and it behaves predictably. If something says it returns a value, it has to return a value, or a default value must be defined explicitly.

So TL;DR: I would like that everything always is required, and that `gomplate` refuses to continue if I don't provide values it needs. The default in `go`s `text/template` is to not error as can be seen [here](https://golang.org/pkg/text/template/#Template.Option), but there you have a full blown programming language to take care of that logic, whereas `gomplate` is something I use because I don't want to write a program to just put variables in a text file, and I can't (or don't want to) check that every value is present. `text/template` doesn't have a `default` function ([ref SO](https://stackoverflow.com/questions/44532017/how-can-i-add-a-default-value-to-a-go-text-template)) for the same reason, but I think it could be useful in gomplate.

----------

I didn't mean to derail the discussion here. And apologies that this got so long. I'm starting to wonder whether we should collect use cases for `gomplate` to help us inform decisions like these.

**Comment by hairyhenderson** [comment]
> I get the feeling that you're not familiar with how `required` works with helm

Not _really_ 😉 - I only just started writing helm charts a few months ago and have never really found a use-case for `required`.

I've had a look at [`required`'s implementation in helm](https://github.com/kubernetes/helm/blob/dede3ff1e1739825e2a039f93acd7cba97c9cd92/pkg/engine/engine.go#L156-L164), and it's a lot simpler than I thought it'd be.

My initial impression is that it could handle the case with `{{ required "..." .Values.foo.bar }}`, where `.Values.foo` is `nil`. But it doesn't - it blows up with a very un-userfriendly error:

```
... at <.Values.foo.bar>: can't evaluate field bar in type interface {}
```

Really it's just checking to see if the value is non-empty and non-`nil`.

This is pretty straightforward. I'll implement `required` the same way helm does it. I'm not sure what namespace it'll end up in, probably `conv.Required`, though I'll add a `required` alias too.

**Comment by hairyhenderson** [comment]
Ah, there's a hitch. And this should've tipped me off:

> I would like that everything always is required, and that gomplate refuses to continue if I don't provide values it needs.

That's actually already gomplate's behaviour, because [I set the `missingkey=error` option](https://github.com/hairyhenderson/gomplate/blob/master/template.go#L32). That also means that `{{ required "..." .Env.UNSET }}` will fail before even executing `required`:

```console
$ gomplate -i '{{ conv.Required "..." .Env.UNSET }}'         
Error: template: <arg>:1:27: executing "<arg>" at <.Env.UNSET>: map has no entry for key "UNSET"
```

This is basically the same as `helm lint --strict`.

So... in short, this'll require a shift in how gomplate handles missing keys, which I'm not super fond of, since I prefer to be stricter by default. One way to accomplish this could be to add a `--strict=false` flag, but users would have to know to run `gomplate --strict=false` in order for `required` to behave as expected.

I think the substance of this particular issue is:
1. gomplate should complain when users don't provide the right data to render a template
   - _this is already current behaviour due to `missingkey=error`_
2. template authors should be able to give users a slightly more meaningful error than `map has no entry for key ...`
  - _to my knowledge this is impossible without changing `missingkey=error`_

I'm not really sure where to go from here 🤷‍♂️

**Comment by mkjpryor-stfc** [comment]
@hairyhenderson

I would like to benefit from the `_FILE` fallback provided by `getenv`, but still require that *either* `VARIABLE` or `VARIABLE_FILE` are set.

The best I can do now is this:

```
{{- $variable := getenv "VARIABLE" }}
{{- assert "VARIABLE[_FILE] is required" (ne $variable "") }}
{{ $variable }}
```

If we had required as implemented in Helm, I could do this instead:

```
{{ getenv "VARIABLE" | required "VARIABLE[_FILE] is required" }}
```

**Comment by mkjpryor-stfc** [comment]
However, if implemented as in Helm the use of `required` is not restricted to environment variables, which I like.

**Comment by hairyhenderson** [comment]
@mkjpryor-stfc thanks for continuing the conversation!

There's a couple alternative ways that are probably no simpler to what you suggest:

```
{{ if not (getenv "VARIABLE") }}{{ fail "VARIABLE[_FILE] is required" }}{{else}}{{getenv "VARIABLE"}}{{ end }}
```

```
{{ not (not (getenv "VARIABLE")) | assert "VARIABLE[_FILE] is required" }}{{getenv "VARIABLE"}}
```

Or if you need to use it for a bunch of items, you could `define` a template:

```
{{ define "required" }}{{ if not . }}{{ fail "Missing required thing" }}{{ else }}{{ . }}{{ end }}{{ end -}}
...
{{ template "required" (getenv "VARIABLE") }}
{{ template "required" (getenv "VARIABLE2") }}
{{ template "required" (getenv "VARIABLE3") }}
```

Neither of these are very tidy.

I just don't think it's possible to implement Helm's `required` without changing the `missingkey=error` option, which TBH I'm not really prepared to do, since it's a major departure from the way gomplate already behaves.

What I _think_ a gomplate-flavoured `required` might look like could simply be - "pass through the given value, _if_ it's non-empty, non-zero, and non-`nil`". 

Returning to @osterman's original request (with added emphasis):

> A `required` function that allows you to declare a particular values entry as required for template rendering. If **the value is empty**, the template rendering will fail with a user submitted error message.

While a `required` function in gomplate will probably never be able to handle cases where the value is _missing_, handling an _empty_ value is actually not so difficult.

Does that sound OK, @osterman, @mkjpryor-stfc, @nicolaiskogheim, and anyone else?

If so, I can certainly try to get that in to v2.8.0.

**Comment by mkjpryor-stfc** [comment]
@hairyhenderson 

I don't particularly like any solution that involves repeating the variable name (potential for mistakes). The template definition is better, but it would be a pain to repeat it for every template file. If `gomplate` had a command-line argument to specify an "environment" file with common template definitions, that could be a goer (perhaps there's another feature request there!).

As I said in my previous comment, I actually like the existing behaviour of `.Env` and `getenv`. What I was envisaging is a `required` function that behaves exactly as you suggested, i.e.:

> What I think a gomplate-flavoured required might look like could simply be - "pass through the given value, if it's non-empty, non-zero, and non-nil".

This would allow me to do what I proposed, i.e.:

```
{{ getenv "VARIABLE" | required "VARIABLE[_FILE] is required" }}
```

So consider this a +1 for that implementation. 👍

**Comment by hairyhenderson** [comment]
Thanks for the feedback @mkjpryor-stfc! I'll wait a bit for @osterman to chime in if this doesn't work, otherwise it's on _the list_...

**Comment by osterman** [comment]
Sorry guys! I missed the discussion (will check my notification settings!)

## use-case 1

We use `gomplate` to render [200+ READMEs](https://github.com/cloudposse) for our GitHub repos. We use [this template](https://raw.githubusercontent.com/cloudposse/build-harness/master/templates/README.md) to process a [YAML spec we created](https://github.com/cloudposse/build-harness/blob/master/README.yaml) to define the parts of a well-formed README. Most parts are optional and we use conditionals. Some parts are required. Rather than emit a "technical exception", we'd like to emit a friendly error message. 

I like the pipeline syntax proposed above.

```
{{- (ds "config").license | required "The `license` field is required" -}}
```

## use-case 2

We use `gomplate` to render our [kops manifest](https://github.com/cloudposse/geodesic/blob/master/rootfs/templates/kops/default.yaml). Most, but not all ENVs are required. I would like to emit a friendly error message with the missing env and what an example value should look like.

```
{{- getenv "KOPS_CLUSTER_NAME" | required "KOPS_CLUSTER_NAME required; e.g. KOPS_CLUSTER_NAME=example.cloudposse.com" -}}
```

--------

Hope that helps clear things up! Thanks @hairyhenderson for waiting on a response. @darend brought this to my attention and is anxiously waiting to use the feature as well.

**Comment by hairyhenderson** [comment]
@osterman thanks for following up! I love hearing that Cloud Posse is using gomplate so effectively! 😁

I'll resurrect my branch and adjust it to implement this.

**Comment by osterman** [comment]
Thanks @hairyhenderson for the quick turnaround

**Comment by mkjpryor-stfc** [comment]
@hairyhenderson Thanks!

**Comment by hairyhenderson** [comment]
🙇
